crates/lumina_server/src/game.rs

```diff
@@ -97,6 +99,77 @@ fn respawn_spaceships(
     }
 }
 
+#[derive(Event)]
+pub struct ResetSpaceships;
+
+/// Resets all spaceship health, energy, weapon and abilities when a game starts
+fn reset_spaceships(
+    mut commands: Commands,
+    mut q_spaceships: Query<
+        (
+            &mut Health,
+            &MaxHealth,
+            &mut Energy,
+            &Spaceship,
+            &PlayerId,
+            Entity,
+            Option<&mut AbilityCooldown>,
+            Option<&ShadowAbilityConfig>,
+            Option<&HealAbilityConfig>,
+        ),
+        With<SourceEntity>,
+    >,
+    q_dash_cooldowns: Query<Entity, With<DashCooldown>>,
+    mut evr_reset_spaceships: EventReader<ResetSpaceships>,
```

Comment:
```suggestion
fn reset_spaceships(
    trigger: Trigger<ResetSpaceships>,
    mut commands: Commands,
    mut q_spaceships: Query<
        (
            &mut Health,
            &MaxHealth,
            &mut Energy,
            &Spaceship,
            &PlayerId,
            Entity,
            Option<&mut AbilityCooldown>,
            Option<&ShadowAbilityConfig>,
            Option<&HealAbilityConfig>,
        ),
        With<SourceEntity>,
    >,
    q_dash_cooldowns: Query<Entity, With<DashCooldown>>,
```

Can try use observer pattern.

Also, I think as an optimization, u can get all the spaceship entities from the `PlayerInfos` thingi by providing the client id which can be obtained from the `Lobby` component. (See the other comment on why we need this)

Reply:
An additional technical challenge bonus: you can try to "parameterize" this system using generics for the abilities.
Maybe split into reset health
and reset ability

so your system can be smaller in size